R/createYaml.R

```diff
@@ -0,0 +1,181 @@
+#' Creates a yaml object storing the parameters that are used for creating the PTXQC report 
+#' and returns these parameters as well as a list of available qc-Metrics objects.
+#' 
+#' Valid parameters are: 
+#' 
+#'    param_useMQPAR, add_fs_col, id_rate_bad, id_rate_great , pg_ratioLabIncThresh , param_PG_intThresh,
+#'    param_EV_protThresh , param_EV_intThresh, param_EV_pepThresh , yaml_contaminants, param_EV_MatchingTolerance,
+#'    param_evd_mbr , param_EV_PrecursorTolPPM, param_EV_PrecursorOutOfCalSD , param_EV_PrecursorTolPPMmainSearch, 
+#'    param_MSMSScans_ionInjThresh, param_OutputFormats and param_PageNumbers 
+#'    
+#'    Please provide them as a list() of this format: list$parameter_name
+#'
+#'
+#' @param yc A yaml class object created by YAMLClass$new()
+#' @param param list of parameters sorted by names; if NULL, will be populated with defaults
+#' @param DEBUG_PTXQC default FALSE
+#' @param MZTAB_MODE default FALSE 
+#' @param txt_files list of paths to MaxQuant files
+#' @param metrics list of metric names that should be plotted; if NULL, will be populated with defaults
+#' @return filled yaml class object, list of parameters used for creating the report and list of qc-Metrics objects
+#' @export
+#'
+#'
+createYaml <- function(yc, param = list(), DEBUG_PTXQC = FALSE, MZTAB_MODE = FALSE, txt_files = NULL, metrics = NULL){
+
+  ##
+  ## YAML default config
+  ##
+  default_param <- list()
+  default_param$param_useMQPAR <- TRUE
+  default_param$add_fs_col <- 14 
+  default_param$id_rate_bad <- 20
+  default_param$id_rate_great <- 35
+  default_param$pg_ratioLabIncThresh <- 4
+  default_param$param_PG_intThresh <- 25
+  default_param$param_EV_protThresh <- 3500
+  default_param$param_EV_intThresh <- 23
+  default_param$param_EV_pepThresh <- 15000
+  default_param$yaml_contaminants <- list("cont_MYCO" = c(name="MYCOPLASMA", threshold=1)) # name (FASTA), threshold for % of unique peptides
+  default_param$param_EV_MatchingTolerance <- 1
+  default_param$param_evd_mbr <- "auto"
+  default_param$param_EV_PrecursorTolPPM <- 20
+  default_param$param_EV_PrecursorOutOfCalSD <- 2
+  default_param$param_EV_PrecursorTolPPMmainSearch <- NA
+  default_param$param_MSMSScans_ionInjThresh <- 10
+  default_param$param_OutputFormats <- c("html", "plainPDF")
+  default_param$param_PageNumbers <- "on"
+  
+  
+  ##
+  ##add missing parameters from default parameter list
+  ##
+  for(i in c(1:length(default_param))){
+    if(!names(default_param)[i] %in% names(param)) param[names(default_param)[i]] <- default_param[i]
+  }
+
+  ##
+  ## check for invalid parameters
+  ##
+  for(i in c(1:length(param))){
+    if(!names(param)[i] %in% c(NA, names(default_param))) {
+      warning(paste0("Invalid parameter detected and removed: ", names(param)[i]))
+      param <- param[-i] 

```

Comment:
this does not work, because removing the i'th element will shift the remaining ones to left and thus the next iteration will skip the next element:
see
```
param = 1:10
names(param) = 1:10

for(i in c(1:length(param))){
  if(!names(param)[i] %in% 22:33) {
    warning(paste0("Invalid parameter detected and removed: ", names(param)[i]))
    param <- param[-i] 
  }
}
```
yields
```
param
 2  4  6  8 10 
 2  4  6  8 10 
```
(but param should be empty)

Reply:
write a test for this to make sure it works